src/Common/src/System/Security/Cryptography/KeyFormatHelper.cs

```diff
@@ -0,0 +1,620 @@
+                    decryptedMemory,
+                    keyReader,
+                    out int innerRead,
+                    out ret);
+
+                if (innerRead != decryptedMemory.Length)
+                {
+                    ret = default;
+                    throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+                }
+
+                bytesRead = read;
+            }
+            catch (CryptographicException e)
+            {
+                throw new CryptographicException(SR.Cryptography_Pkcs8_EncryptedReadFailed, e);
+            }
+            finally
+            {
+                CryptographicOperations.ZeroMemory(decryptedMemory.Span);
+                ArrayPool<byte>.Shared.Return(decrypted);
+            }
+        }
+
+        internal static AsnWriter WritePkcs8(AsnWriter algorithmIdentifierWriter, AsnWriter privateKeyWriter)
+        {
+            // Ensure both input writers are balanced.
+            ReadOnlySpan<byte> algorithmIdentifier = algorithmIdentifierWriter.EncodeAsSpan();
+            ReadOnlySpan<byte> privateKey = privateKeyWriter.EncodeAsSpan();
+
+            Debug.Assert(algorithmIdentifier.Length > 0, "algorithmIdentifier was empty");
+            Debug.Assert(algorithmIdentifier[0] == 0x30, "algorithmIdentifier is not a constructed sequence");
+            Debug.Assert(privateKey.Length > 0, "privateKey was empty");
+
+            // https://tools.ietf.org/html/rfc5208#section-5
+            //
+            // PrivateKeyInfo ::= SEQUENCE {
+            //   version                   Version,
+            //   privateKeyAlgorithm       PrivateKeyAlgorithmIdentifier,
+            //   privateKey                PrivateKey,
+            //   attributes           [0]  IMPLICIT Attributes OPTIONAL }
+            // 
+            // Version ::= INTEGER
+            // PrivateKeyAlgorithmIdentifier ::= AlgorithmIdentifier
+            // PrivateKey ::= OCTET STRING
+            AsnWriter writer = new AsnWriter(AsnEncodingRules.DER);
+
+            // PrivateKeyInfo
+            writer.PushSequence();
+
+            // https://tools.ietf.org/html/rfc5208#section-5 says the current version is 0.
+            writer.WriteInteger(0);
+
+            // PKI.Algorithm (AlgorithmIdentifier)
+            WriteEncodedSpan(writer, algorithmIdentifier);
+            
+            // PKI.privateKey
+            writer.WriteOctetString(privateKey);
+
+            // We don't currently accept attributes, so... done.
+            writer.PopSequence();
+            return writer;
+        }
+
+        internal static unsafe AsnWriter WriteEncryptedPkcs8(
+            ReadOnlySpan<char> password,
+            AsnWriter pkcs8Writer,
+            PbeParameters pbeParameters)
+        {
+            System.Text.Encoding encoding = System.Text.Encoding.UTF8;
+            int requiredBytes = encoding.GetByteCount(password);
+            Span<byte> passwordBytes = stackalloc byte[0];
```

Comment:
This is to work around a compiler error about span lifetimes?

Reply:
Yep